vertx-web-client/src/main/java/io/vertx/ext/web/client/impl/cache/CacheInterceptor.java

```diff
@@ -0,0 +1,228 @@
+    if (!options.isCachingEnabled()) {
+      context.next();
+      return;
+    }
+
+    switch (context.phase()) {
+      case SEND_REQUEST:
+        handleSendRequest((HttpContext<Buffer>) context);
+        break;
+      case DISPATCH_RESPONSE:
+        handleDispatchResponse((HttpContext<Buffer>) context);
+        break;
+      default:
+        context.next();
+        break;
+    }
+  }
+
+  private void handleSendRequest(HttpContext<Buffer> context) {
+    HttpRequestImpl<Buffer> requestImpl = (HttpRequestImpl<Buffer>) context.request();
+    if (!HttpMethod.GET.equals(requestImpl.method())) {
+      context.next();
+      return;
+    }
+
+    Vary variation = selectVariation(requestImpl);
+    if (variation == null) {
+      context.next();
+      return;
+    }
+
+    cacheStore
+      .get(new CacheKey(context.request(), variation))
+      .compose(cached -> respondFromCache(context, cached))
+      .onComplete(ar -> {
+        if (ar.succeeded()) {
+          context.set(IS_CACHE_DISPATCH, true);
+          context.dispatchResponse(ar.result());
+        } else {
+          context.next();
+        }
+      });
+  }
+
+  private void handleDispatchResponse(HttpContext<Buffer> context) {
+    Boolean isCacheDispatch = context.get(IS_CACHE_DISPATCH);
+    if (isCacheDispatch != null && isCacheDispatch) {
+      context.next();
+      return;
+    }
+
+    CachedHttpResponse responseToValidate = context.get(REVALIDATION_RESPONSE);
+    if (responseToValidate != null) {
+      // We're revalidating a cached response
+      processRevalidationResponse(context, responseToValidate).onComplete(ar -> {
+        if (ar.succeeded()) {
+          context.response(ar.result());
+        }
+        context.next();
+      });
+    } else {
+      // We're storing a new response in cache
+      processResponse(context, null).onComplete(ar -> {
+        // TODO: is this right? if we don't need the data then use a callback, not a future
+        context.next();
+      });
+    }
+  }
+
+  private Vary selectVariation(HttpRequest<?> request) {
+    CacheVariationsKey key = new CacheVariationsKey(request);
+    Set<Vary> possibleVariations = variationsRegistry.getOrDefault(key, Collections.emptySet());
+
+    for (Vary variation : possibleVariations) {
+      if (variation.matchesRequest(request)) {
+        return variation;
+      }
+    }
+
+    return null;
+  }
+
+  private Future<HttpResponse<Buffer>> processResponse(HttpContext<Buffer> context, CachedHttpResponse cachedResponse) {
+    HttpRequest<Buffer> request = context.request();
+    HttpResponse<Buffer> response = context.response();
+
+    if (options.getCachedStatusCodes().contains(response.statusCode())) {
+      // Request was successful, attempt to cache response
+      return cacheResponse(request, response).map(response);
+    } else if (cachedResponse != null && cachedResponse.useStaleIfError()) {
+      // TODO: we should check that this is an error too
```

Comment:
we should check that `context.response()` was an error response before checking the `cachedResponse.useStaleIfError()` flag. If the response didn't result in an error, we should just use that instead of falling back to the cached response.